### PR TITLE
terminal tabs UI

### PR DESCRIPTION
**Pull request by meganrogge**
Part of #120241

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/media/terminal.css [hunk 1]
```diff
@@ -3,6 +3,13 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+
+.tabs-widget-container {
+	width: 20%;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/terminal.css:8]: When the split view works we'll want to remove these explicit widths

File: src/vs/workbench/contrib/terminal/browser/media/terminal.css [hunk 2]
```diff
@@ -3,6 +3,13 @@
...
+	float: left;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/terminal.css:9]: `display: flex` is way better than using float for layouts

File: src/vs/workbench/contrib/terminal/browser/tabsView.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	) {
+		if (context === 'terminal') {
+			this._splitView = new SplitView(container, { orientation: Orientation.HORIZONTAL });
+			this._widget = _instantiationService.createInstance(TerminalTabsWidget, this._splitView.el);
+		}
+	}
+
+	public layout(width: number, height: number): void {
+		this._splitView?.layout(100);
+		this._terminalService.terminalTabs.forEach(t => t.layout(width - 100, height));
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/tabsView.ts:32]: The `TerminalTabsWidget` and terminal tabs container would be `SplitPane` children of the `SplitView`, like how a `TerminalTab`'s `SplitPaneContainer` has a child `SplitPane` for every instance in the tab.

File: .vscode/settings.json [hunk 4]
```diff
@@ -86,5 +86,4 @@
 	},
 	"typescript.tsc.autoDetect": "off",
 	"notebook.experimental.useMarkdownRenderer": true,
-	"testing.autoRun.mode": "onlyPreviouslyRun",
```
Comment (Tyriar) [inline .vscode/settings.json:89]: Accidentally committed?

File: src/vs/workbench/contrib/terminal/browser/tabsView.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+const FIND_FOCUS_CLASS = 'find-focused';
+
+export class TabsView extends Disposable {
+	private _menu: IMenu;
+	_width: number;
+	_height: number;
+	private _cancelContextMenu: boolean = false;
+	private _tabsElement: HTMLElement;
+	private _splitView!: SplitView;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/tabsView.ts:36]: Should we move the split view creation into the ctor to avoid the `!`?

Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/tabsView.ts:36]: readonly?

File: src/vs/workbench/contrib/terminal/browser/tabsView.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+	private readonly _splitViewDisposables = this._register(new DisposableStore());
+	private _children: SplitTabsPane[] = [];
+	private _terminalContainer: HTMLElement;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/tabsView.ts:39]: readonly?

File: src/vs/workbench/contrib/terminal/browser/tabsView.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+		this._createSplitView();
+		_findWidget.focusTracker.onDidFocus(() => this._terminalContainer!.classList.add(FIND_FOCUS_CLASS));
+	}
+
+	public get splitView(): SplitView {
+		return this._splitView;
+	}
+
+	private _createSplitView(): void {
+		if (this._splitView) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/tabsView.ts:76]: We can avoid this I think as it's only created once?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts [hunk 8]
```diff
@@ -230,6 +230,17 @@ export const terminalConfiguration: IConfigurationNode = {
...
+		'terminal.integrated.showTabs': {
+			description: localize('terminal.integrated.showTabs', 'Controls whether or not the terminal tabs widget is shown'),
+			type: 'boolean',
+			default: false
+		},
+		'terminal.integrated.tabsLocation': {
+			'type': 'string',
+			'enum': ['left', 'right'],
+			'default': 'left',
+			'description': localize('sideBarLocation', "Controls the location of the terminal tabs, either left or right of the terminal container.")
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts:242]: Id needs renaming

**Review by Tyriar — CHANGES_REQUESTED** [review]
Reviewing on my phone so I couldn't look into some things myself 🙂

File: src/vs/workbench/contrib/terminal/browser/terminalTabbedView.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+		@IInstantiationService instantiationService: IInstantiationService
+	) {
+		super();
+		this._instantiationService = instantiationService;
+		this._showTabs = terminalService.configHelper.config.showTabs;
+		this.TAB_TREE_INDEX = terminalService.configHelper.config.tabsLocation === 'left' ? 0 : 1;
+		this.TERMINAL_CONTAINER_INDEX = terminalService.configHelper.config.tabsLocation === 'left' ? 1 : 0;
+		this._terminalTabTree = document.createElement('div');
+		this._terminalTabTree.classList.add('tabs-widget');
+		this._tabsWidget = this._instantiationService.createInstance(TerminalTabsWidget, this._terminalTabTree);
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalTabbedView.ts:44]: no, we want to always create it. and set it's visibilty I think

File: src/vs/workbench/contrib/terminal/common/terminal.ts [hunk 10]
```diff
@@ -103,6 +103,8 @@ export interface ITerminalConfiguration {
 	};
 	profiles: ITerminalProfiles;
 	useWslProfiles: boolean;
+	showTabs: boolean;
+	tabsLocation: 'left' | 'right';
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/common/terminal.ts:107]: yes

File: src/vs/workbench/contrib/terminal/browser/terminalTabsWidget.ts [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	rerender(): void {
+		this.setChildren(null, undefined);
+		const children = createTerminalTabsIterator(this._terminalService.terminalTabs);
+		this.setChildren(null, children);
+	}
+}
+
+class TerminalTabsDelegate implements IListVirtualDelegate<TerminalTab> {
+	getHeight(element: any): number {
+		return 24;
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalTabsWidget.ts:87]: https://github.com/microsoft/vscode/blob/10d92e5efaf67a39660260f4fa5e44b3f847c597/src/vs/workbench/contrib/preferences/browser/tocTree.ts#L145-L145

**Review by Tyriar — APPROVED** [review]

**Comment by Tyriar** [comment]
Exception on right click:

Screen Shot 2021-04-08 at 5 44 32 AM

**Pull request merged**